Changes for IITC

modos189 opened the pull request:
On one hand, this change is specific to IITC, and on other hand - for greater similarity with the leaflet. Perhaps we should create a parameter to determine whether the container should be destroyed.
See: https://github.com/IITC-CE/Leaflet.Canvas-Markers/commit/b6c228a1b1752255c6b82be3c91473fc02b578af

modos189 committed:
Fix typo

Commit:
return general Layer features lost in refactoring into Renderer

Original intentions of that refactoring is not clear,
perhaps it was intermediate step before transforming into true Renderer.

But in current state it is not actually Renderer, so we need to revert some parts,
to continue using it as independent Layer.

Notes:
  `CanvasMarkerLayer` is still independent entity which does not fit flawlessly into Leaflet ecosystem.
  It would be much more convenient if instead of `CanvasMarkerLayer` we have `CanvasMarker`,
  able to be used as direct and transparent replacement for `L.Marker`.
  Though currently I am not sure how to do it best:
  - may be we need dedicated Renderer class
  - or may be it can be simpler, and all we need - `CanvasMarker` inherited from `L.Path`.

modos189 committed:
Changing Canvas size when changing screen size

johnd0e commented:
> https://github.com/IITC-CE/Leaflet.Canvas-Markers/pull/1/commits/a52cbaf383b38e1c642bc5d4e34ae399688e8ba5

Nice job!

modos189 committed:
Preparation to support options.padding

modos189 commented:
Maybe you'll be able to finish options.padding support. Ornaments are constantly shifting: either by moving the map or changing the screen size

johnd0e commented:
> Ornaments are constantly shifting

We need padding to solve this, or this issue caused by unfinished padding support?

modos189 commented:
It's not ready. Maybe I should have sent it to a separate branch.
Padding is needed to draw portals that are outside the visible part of the screen, but close to it, just like in Leaflet and IITC. I have implemented this, but there are some conflicts when moving the map. Ornaments move more than portals.

johnd0e commented:
As this feature is not blocker it makes sense to put it separate branch.

And tell me more about 419b1ab: what is 'atypical use'?

modos189 committed:
Correction marker positions when panning or resizing a window.

modos189 commented:
I was able to fix it. Padding works correctly.

>  what is 'atypical use'?

Our code tries to remove those markers which may not exist.

https://github.com/IITC-CE/ingress-intel-total-conversion/blob/256f9c53a866a254b40e40b1f766e58b4d2a89f4/code/ornaments.js#L69-L73

That's why I added additional checks to correct the errors.

johnd0e commented on src/plugin/leaflet.canvas-markers.js:
```diff
@@ -283,8 +283,8 @@ function layerFactory(L) {
-                pointPos.x - marker.options.icon.options.iconAnchor[0]-(this._topLeftOffset?this._topLeftOffset.x:0),
-                pointPos.y - marker.options.icon.options.iconAnchor[1]-(this._topLeftOffset?this._topLeftOffset.y:0),
+                pointPos.x - marker.options.icon.options.iconAnchor[0]-this._map._mapPane._leaflet_pos.x-(this._topLeftOffset?this._topLeftOffset.x:0),
+                pointPos.y - marker.options.icon.options.iconAnchor[1]-this._map._mapPane._leaflet_pos.y-(this._topLeftOffset?this._topLeftOffset.y:0),
```
Here we are deep into leaflet internals.
May be we could use [containerPointToLayerPoint](https://leafletjs.com/reference-1.4.0.html#map-containerpointtolayerpoint)/[layerPointToContainerPoint](https://leafletjs.com/reference-1.4.0.html#map-layerpointtocontainerpoint) instead?

johnd0e commented:
> I was able to fix it. Padding works correctly.

Great!
But check upper comment please ^^^:

johnd0e commented on src/plugin/leaflet.canvas-markers.js:
```diff
@@ -3,6 +3,13 @@
+            // e.g. 0.1 would be 10% of map view in each direction
+            padding: 0.1
+        },
+
```
Shouldn't we inherit this value from map renderer?

modos189 replied:
In Leaflet `padding` value is specified in `Renderer` class (`var Renderer = Layer.extend({`) and is not overridden.
It would be useful to assign an indent equal to that in `window.map` in iitc, but this is hardly possible.

johnd0e replied:
> It would be useful to assign an indent equal to that in window.map in iitc, but this is hardly possible.

Why hardly? I suppose we just need `map._renderer.options.padding`.
Or `map.getRenderer({options:{}}).options.padding`.

Then in our `boot.js` we can do like this:
```boot.js
L.CanvasIconLayer.mergeOptions({
  padding: map._renderer.options.padding`
});
```

johnd0e commented on src/plugin/leaflet.canvas-markers.js:
```diff
@@ -24,7 +24,8 @@ function layerFactory(L) {
         onRemove: function () {
-            this._destroyContainer();
+            // TODO
+            // this._destroyContainer();
```
> Call this._destroyContainer() is unique to this plugin and redundant

Currently `L.CanvasMarkerLayer` has dual nature:
- `L.Layer`, from which it's inherited initially.
- And after https://github.com/Spaction/Leaflet.Canvas-Markers/commit/153525e045e1c4ec22a0ecb02d965366e99852a0 it has some features of `L.Renderer` (but still is not inherited from it)

I can guess that mentioned commit was just a step to make true renderer which can be used transparently (#2).
But in it's current (perhaps unfinished) state `L.CanvasMarkerLayer` is rather layer, so it should'n destroy it's container on remove.

I wonder if we've chosen right fork to continue..

Commit:
fix broken removeMarker

johnd0e commented:
> Our code tries to remove those markers which may not exist.

`removeMarker` [was](https://github.com/IITC-CE/Leaflet.Canvas-Markers/commit/eda858e81a3df5c62cda7de164b4dc99e243edf5) broken anyway.

Also I am going to fix our code, so we do not need additional checks in upstream.

Although, those checks [may make sense](https://github.com/IITC-CE/Leaflet.Canvas-Markers/issues/10) in general (but not in this branch).

johnd0e commented on src/plugin/leaflet.canvas-markers.js:
```diff
@@ -118,7 +124,7 @@ function layerFactory(L) {
             if (this._map._animatingZoom && this._bounds) { return; }
 
-            var p = 0,
+            var p = options.padding,
```
```suggestion
            var p = this.options.padding,
```
Bug here:

modos189 replied:
I agree

johnd0e commented on src/plugin/leaflet.canvas-markers.js:
```diff
@@ -50,17 +59,12 @@ function layerFactory(L) {
             var scale = this._map.getZoomScale(zoom, this._zoom),
                 position = L.DomUtil.getPosition(this._container),
-                viewHalf = this._map.getSize().multiplyBy(0.5),
+                viewHalf = this._map.getSize().multiplyBy(0.5+options.padding),
```
```suggestion
                viewHalf = this._map.getSize().multiplyBy(0.5 + this.options.padding),
```

modos189 replied:
I agree

Commit:
fix exception in mouse event handler with empty CanvasMarkerLayer

`this._markers` is undefined in `_searchPoints`, called from event listeners.
Now we check markers existence in the beginning of handlers.

Note 1:
Could be fixed in some another place:
 e.g. check `_markers` in `_searchPoints`
 ...or initialize (empty) `_markers` on CanvasMarkerLayer init

Note 2:
`!this._map` removed from condition in `_onMouseMove:` 'cause
it's never should be true as we detach listeners on remove from map

johnd0e commented:
@modos189 
I am going to clean up and rewrite this branch, tell me when you ready.
[Separate branches](https://github.com/IITC-CE/Leaflet.Canvas-Markers/branches) for all features already created (todo: a52cbaf383b38e1c642bc5d4e34ae399688e8ba5)

johnd0e commented on src/plugin/leaflet.canvas-markers.js:
```diff
@@ -101,8 +101,6 @@ function layerFactory(L) {
             this._update();
             this._updateTransform(this._center, this._zoom);
-
-            this._update();
```
Is this change connected with a52cbaf383b38e1c642bc5d4e34ae399688e8ba5?

johnd0e commented:
Please check comments regarding padding option.
I have implemented fixes in [`padding-option`](https://github.com/IITC-CE/Leaflet.Canvas-Markers/tree/padding-option) branch, but not tested yet.

modos189 commented on src/plugin/leaflet.canvas-markers.js:
```diff
@@ -274,14 +274,18 @@ function layerFactory(L) {
-            L.Util.splitWords(event).forEach(function (e) {
-                self._userEvents[e] = func;
+            L.Util.splitWords(types).forEach(function (type) {
+                if (internal.indexOf(event) === -1) {
```
`event` is not defined

modos189 commented on src/plugin/leaflet.canvas-markers.js:
```diff
@@ -274,14 +274,18 @@ function layerFactory(L) {
+                if (internal.indexOf(event) === -1) {
+                    L.Evented.prototype._on.call(self, type, fn, context);
+                } else {
+                    self._userEvents[e] = fn;
```
`e` is not defined

johnd0e commented:
My fault, thank you for pointing out.

modos189 committed:
Simplification of coordinate calculation

modos189 commented:
I saw you similar change in `padding-option` branch. But there is an error: instead of `layerPointToContainerPoint` you should use `containerPointToLayerPoint`

johnd0e commented:
> I saw you similar change in `padding-option` branch

I've just updated that branch with your last commit.
Let's continue padding-related development there.

johnd0e commented:
As all needed for IITC features are now separated between own [branches](https://github.com/IITC-CE/Leaflet.Canvas-Markers/branches), all we need - merge them together, like this:
```
git checkout master
git checkout -b changes-for-IITC
git merge --no-ff fixes canvas-resize layer-features-back padding-option
git push --force
```

I've already done this in [`test`](https://github.com/IITC-CE/Leaflet.Canvas-Markers/tree/test) branch, and about to rewrite `changes-for-IITC` branch with the content of `test`.

All future updates should go to feature branches, here we'll need merge them together only.

P.S.
And I suppose we need to open corresponding PR's in Spaction's repo.

johnd0e commented:
@modos189 
I finished with all branches except [`padding-option`](https://github.com/IITC-CE/Leaflet.Canvas-Markers/tree/padding-option).

I am unable to check it, as I just do not understand how it should function.
Let's start from related issue: https://github.com/IITC-CE/ingress-intel-total-conversion/issues/184

Commit:
fixup! Preparation to support options.padding

Commit:
removeMarker: fix exception when CanvasMarkerLayer is not attached to map

Commit:
fixup! fixup! Preparation to support options.padding

Commit:
fix exception if image is not loaded (yet) / bad url / etc

Was protected in marker add time, but not on zoom/pan

Sample:
```
VM2992:16252 Uncaught DOMException: Failed to execute 'drawImage' on 'CanvasRenderingContext2D': The HTMLImageElement provided is in the 'broken' state.
    at NewClass._drawImage (<anonymous>:16252:23)
    at NewClass._drawMarker (<anonymous>:16239:22)
    at <anonymous>:16200:22
    at Array.forEach (<anonymous>)
    at NewClass._draw (<anonymous>:16179:16)
    at NewClass._redraw (<anonymous>:16077:18)
    at NewClass.fire (<anonymous>:1867:11)
    at NewClass._move (<anonymous>:5506:9)
    at NewClass._onZoomTransitionEnd (<anonymous>:5963:8)
```

Commit:
fixup! fixup! fixup! Preparation to support options.padding

Commit:
fix to expose L.CanvasIconLayer class

This was originally meant but missed (see _full.js / _standalone.js)

Commit:
Merge branches 'fixes', 'canvas-resize', 'layer-features-back' and 'padding-option' into test

Commit:
fix https://github.com/IITC-CE/Leaflet.Canvas-Markers/issues/11

modos189 commented:
It's all working now.

johnd0e commented:
Will actually see on next event)

But please take attention to minor issues left in that repo.

johnd0e commented:
@modos189
I am not sure that 2 latest commits are needed.
See test branch (as I said in previous message)

modos189 commented:
In https://github.com/IITC-CE/ingress-intel-total-conversion/pull/181 which branch is being used? I used that version and I thought it was from test branch

up: Indeed, these changes already exist in the test branch

johnd0e commented:
Yeah, and it already has upper 2 fixes, isn't it?

And `test` branch itself is merged from others our branches (except PR's subject), so I wonder why do we need these doubled commits again.

johnd0e commented:
Do you still need this branch? If not then I will rewrite it with actual commits.

modos189 commented:
No, you can rewrite

johnd0e commented:
Please check this comment: https://github.com/IITC-CE/Leaflet.Canvas-Markers/commit/71eab756484bf298384f5059f15d2fd86fb0b182#commitcomment-33502466

modos189 commented:
I'm totally confused in branches, so I created a new branch for onresize method: https://github.com/IITC-CE/Leaflet.Canvas-Markers/tree/onresize

Merge after this: https://github.com/IITC-CE/Leaflet.Canvas-Markers/pull/1#issuecomment-489002729

johnd0e commented:
> I'm totally confused in branches

We have independent branch for every new feature here. We keep it so it will be easy to make PR's to parent repo in the future.

And all that branches merge together here, in `changes-for-IITC` - it is not independent branch.

So to complete our work here we should finish last branch, for what we have concerns: `padding-option`.
I think it is not ok, check comments for f451114f2f74e9cd896cd8a160e3d34bffdcc7ff

modos189 commented:
@johnd0e https://github.com/IITC-CE/Leaflet.Canvas-Markers/commit/1d88dcfee3a83e5ddd3e41abab6ed3f996623bb7

johnd0e commented:
> @johnd0e [1d88dcf](https://github.com/IITC-CE/Leaflet.Canvas-Markers/commit/1d88dcfee3a83e5ddd3e41abab6ed3f996623bb7)

Good! But see question there.

Also question about `_updateTransform` left unanswered in https://github.com/IITC-CE/Leaflet.Canvas-Markers/commit/f451114f2f74e9cd896cd8a160e3d34bffdcc7ff
